R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
+
+  # internal --------------------------
+
+  # The strategy is to split the data frame for each source unit and carry out
+  # conversion separately per data frame. This is the most performant options
+  # since there can only be as many expensive calls to `toUnit()`/`rClr` as
+  # there are source units.
+  #
+  # The problem occurs when source units are missing (`NA`). The `toUnit`
+  # function can handle them but not `split()`, which would drop the entire
+  # section of the data frame corresponding to `NA` and thus there will be loss
+  # of data when a data frame is split into a list of data frames.
+  #
+  # The trick is to create copies of source unit and molecular weight columns
+  # and fill in the missing values with something other than `NA`. Using these
+  # new columns with `split()` makes sure that parts of a data frame where
+  # source units are missing won't be dropped. Note that the original columns
+  # containing source units remain untouched.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
+        .fns = as.character,
+        .names = "{.col}Split"
+      )
+    ) %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Split$"),
+        .fns = ~ tidyr::replace_na(.x, "missing")
+      )
+    ) %>%
+    # `split()` will change the row order of the data frame depending on the
+    # alphabetical order of the levels of the variable used to split the data
+    # frame into a list.
+    #
+    # Therefore, an internal row identifier is kept to restore the original
+    # data frame row order before data is returned.
+    dplyr::mutate(.rowidInternal = dplyr::row_number())
+
+  # xUnit --------------------------
+
+  xDataList <- .removeEmptyDataFrame(split(data, data$xUnitSplit))
+  data <- dplyr::bind_rows(lapply(xDataList, .xUnitConverter, xTargetUnit))
+
+  # yUnit ----------------
+
+  yDataList <- .removeEmptyDataFrame(split(data, list(data$yUnitSplit, data$molWeightSplit)))
+  data <- dplyr::bind_rows(lapply(yDataList, .yUnitConverter, yTargetUnit))
+
+  # yUnit error ----------------
+
+  if ("yErrorValues" %in% names(data)) {
+    yErrorDataList <- .removeEmptyDataFrame(split(data, list(data$yErrorUnitSplit, data$molWeightSplit)))
+    data <- dplyr::bind_rows(lapply(yErrorDataList, .yErrorUnitConverter, yTargetUnit))
+  }
+
+  # Restore the original row order using the internal row id
+  data <- dplyr::arrange(data, .rowidInternal)
+
+  # Remove all columns that were added only for internal workings of the function.
+  data <- dplyr::select(data, -dplyr::matches("Split$|.rowidInternal"))
+
+  return(data)
+}
+
+#' Remove empty data frames sometimes produced due to the non-existent
+#' combination of source unit and molecular unit.
+#'
+#' @keywords internal
+#' @noRd
+.removeEmptyDataFrame <- function(x) {
```

Comment:
I don't like this method because it seems to have side effects and also returns something
I cannot tell if we are modifying x in place or not. 
Also what is x?

Reply:
>  it seems to have side effects

What's the side effect here?

> I cannot tell if we are modifying x in place or not.

**Only** environments exhibit reference semantics in R, everything else has a value semantics. 

So, no, we are not modifying `x` in place. We are creating a new copy and binding it to the same name.

Reply:
And, `x` is a list.

Reply:
>Only environments exhibit reference semantics in R

R6 class also. They are maybe environment, maybe not. I don't know

Reply:
well it looks like you are modifying x
x <- purrr::keep(x, ~ nrow(.x) > 0L)
  x
if not, just return `purrr::keep(x, ~ nrow(.x) > 0L)` and name that parameter correctly unless it is common in R to name parameter x for non inline method (e.g. lambda expression)

Reply:
Parameter `x` is idiomatic for atomic vector or list arguments.

E.g.

``` r
formals(list2DF)
#> $x
#> list()
#> 
#> $nrow
#> NULL
```

<sup>Created on 2022-04-22 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

But happy to rename it to something else.